### PR TITLE
Add mechanism to auto update most of packages

### PR DESCRIPTION
**Pull request by MrAdityaAlok**
- [x] Add package neovim-nightly (initially PR was for this)
- [x] Add functions to override auto-update script. Suggested by @thunder-coding
- [ ] Update wiki
- [x] Restrict build scripts from using `GITHUB_API_TOKEN`. No need now since wirkflow only run for already present packages, not also for PR.

#### Main function (overridable)

*This should be overrided for using custom update method. For example [look here](https://github.com/termux/termux-packages/blob/3867cbacd5172511048ac2e3a5d228783af253fa/packages/neovim-nightly/build.sh#L27).*

- [x] `termux_pkg_auto_update` : determine which update script to use for auto-updating

#### Utility/API (available publicly)

- [x] `termux_pkg_upgrade_version`: to write new version to build.sh
- [x] `termux_pkg_is_update_nedded`: determine wheteher package needs to be updated
- [x] `termux_pkg_get_last_updated`: get last updated date for github housed packages. It is used for api cache call to lower api call rate.
- [x] `termux_github_api_get_tag`: returns latest/newest version tag
- [x] `termux_gitlab_api_get_tag`: returns latest/newest version tag
- [x] `termux_repology_api_get_latest_version`: returns latest version

#### Internal functions (should not be overrided)

- [x] `termux_github_auto_update` : default update algorithm for packages hosted on github.com
- [x] `termux_gitlab_auto_update`: default update algorithm for pkgs hosted on gitlab.com
- [x] `termux_repology_auto_update`: default update algorithm for pkgs hosted on repology.org

#### Tests
- [x] Test new packages for auto-update: it can be tested locally as `./update_packages <package name>`  (`GITHUB_TOKEN` should be set in env.)
- Test existing packages (I will test and enable update for existing packages in another PR).

### CODE REVIEW

File: .github/workflows/package_updates.yml
Comment (thunder-coding) [inline .github/workflows/package_updates.yml:30]: Instead of doing something like this, why not support having a `termux_package_latest_version` function in `build.sh` which can be used to fetch the latest version. If this is implemented, some other packages can also be auto-updated in a similar way

**Review by thunder-coding — CHANGES_REQUESTED** [review]

File: .github/workflows/package_updates.yml
Comment (MrAdityaAlok) [reply .github/workflows/package_updates.yml:30]: Yes, I would like to support. But could you explain more about how it would be used? I need some more context to work on it.

Comment (MrAdityaAlok) [reply .github/workflows/package_updates.yml:30]: I'm working on it.

Comment (MrAdityaAlok) [reply .github/workflows/package_updates.yml]: @thunder-coding have a look.

File: packages/neovim-nightly/src-nvim-os-stdpaths.c.patch
Comment (thunder-coding) [inline packages/neovim-nightly/src-nvim-os-stdpaths.c.patch:11]: Shouldn't this be `@TERMUX_HOME@/.local/share/:@TERMUX_PREFIX@/share/` ?

File: scripts/bin/auto-update/utility.sh
Comment (thunder-coding) [inline scripts/bin/auto-update/utility.sh:1]: Wouldn't it make sense to move this to the `scripts/updates` directory?

**Review by thunder-coding — CHANGES_REQUESTED** [review]

File: packages/neovim-nightly/src-nvim-os-stdpaths.c.patch
Comment (thunder-coding) [reply packages/neovim-nightly/src-nvim-os-stdpaths.c.patch:11]: My bad the same patch is for neovim, so resolving this also it seems to respect `~/.local/share`

File: scripts/bin/auto-update/utility.sh
Comment (MrAdityaAlok) [reply scripts/bin/auto-update/utility.sh]: Done

File: scripts/updates/utility.sh
Comment (thunder-coding) [inline scripts/updates/utility.sh:1]: Consider moving this to `scripts/updates/termux_pkg_upgrade_version`

File: scripts/updates/utility.sh
Comment (thunder-coding) [inline scripts/updates/utility.sh:56]: Rename function to `termux_github_auto_update` and move to `scripts/updates/termux_github_auto_update.sh`

This is because after this get's merged, I have plans to work on `termux_repology_auto_update`

**Review by thunder-coding — CHANGES_REQUESTED** [review]
Some final changes

After this, I will experiment with these to make sure they don't cause troubles on `master`

**Comment by thunder-coding** [comment]
> Update wiki

Leave this on me, I will do this after merging

> Hide GITHUB_API_TOKEN from auto-update scripts.

Glad you are looking into it :)

File: scripts/updates/utility.sh
Comment (MrAdityaAlok) [reply scripts/updates/utility.sh:56]: What does name `termux_github_auto_update` imply? Does this function should only auto update projects hosted on github?

Comment (thunder-coding) [reply scripts/updates/utility.sh:56]: I feel it would make maintaining it easier if GitHub, Repology (to be added in future) and other update scripts should be kept at different places. The default `termux_pkg_auto_update` will just do the work of choosing between what update script to use depending on `$TERMUX_PKG_SRCURL`

**Comment by MrAdityaAlok** [comment]
Do not review, I am still working on it..

File: scripts/updates/utility.sh
Comment (MrAdityaAlok) [reply scripts/updates/utility.sh]: @thunder-coding if you do not mind, I would like to work on `termux_repology_auto_update`. If you are fine with it then please guide me what would you do for repology update...

Comment (thunder-coding) [reply scripts/updates/utility.sh]: Repology does have an API too, did you check out its docs. Probably there's no reason to scrap the website.
https://repology.org/api

Comment (MrAdityaAlok) [reply scripts/updates/utility.sh]: That doesn't provide information per repository basis, so I decided to write my own scrapper.

**Comment by MrAdityaAlok** [comment]
This should now be ready for review.

**Comment by Ladla-saith** [comment]
Plz tell me how to grow my  commands

**Comment by MrAdityaAlok** [comment]
This should now be ready for review.

**Comment by MrAdityaAlok** [comment]
Please ignore ci failures, it is intended (due to lack of permission to access GH_API_KEY for PRs).

**Comment by buttaface** [comment]
Anything holding this up? [We're down to only 73% of our packages being up to date](https://repology.org/repository/termux), we used to be in the 90s a year or two ago.

**Comment by MrAdityaAlok** [comment]
@buttaface I need to review it again. 
Initially I didn't see @termux maintainers interest in this PR,  so I moved on to work on #9333 . 
But now this PR should be ready for review in 2-3 days.

**Comment by buttaface** [comment]
No hurry, but I think we should definitely expand beyond only auto-updating github-housed projects.

**Comment by MrAdityaAlok** [comment]
Please ignore ci failures. They are due to not upto date `neovim-nightly`'s checksum. Just ensure to run `./srcipts/bin/update-packages neovim-nightly` before merging to update its checksum or ping me to do so.

**Comment by MrAdityaAlok** [comment]
Ping @Grimler91

**Comment by Grimler91** [comment]
@MrAdityaAlok commits and structure looks reasonable, thanks. 

Please update the nightly hashsum so that we can see if the build succeeds, and so that debs can be inspected if needed

File: scripts/updates/utils/termux_pkg_is_update_needed.sh
Comment (Grimler91) [inline scripts/updates/utils/termux_pkg_is_update_needed.sh:64]: We have a script [scripts/bin/apt-compare-versions](https://github.com/termux/termux-packages/blob/master/scripts/bin/apt-compare-versions) which aims at doing pretty much the same thing as this script, with the difference that it prints the result in a human-friendly way. Please remove the other one

Comment (MrAdityaAlok) [reply scripts/updates/utils/termux_pkg_is_update_needed.sh:64]: Should I remove this whole function or only the code making it executable?

Comment (Grimler91) [reply scripts/updates/utils/termux_pkg_is_update_needed.sh:64]: This looks like a more rigorous implementation than the other one, so I suggest to use this one and remove scripts/bin/apt-compare-versions. Possible improvement to this termux_pkg_is_update_needed.sh script would be to add a "human-readable" mode, which prints the versions as the other script did, but since that isn't related to auto updates or neovim it's not something that has to be added as part of this PR IMO

Comment (MrAdityaAlok) [reply scripts/updates/utils/termux_pkg_is_update_needed.sh:64]: Still I would add that functionality in this PR only. I have exams in upcoming days, so would be busy there for few weeks...

Comment (Grimler91) [reply scripts/updates/utils/termux_pkg_is_update_needed.sh:64]: Yeah, that's fine. Fix the email address and I can merge it

Comment (MrAdityaAlok) [reply scripts/updates/utils/termux_pkg_is_update_needed.sh]: I have added that functionality. Please review again.

**Comment by Grimler91** [comment]
Github's UI failed to detect the merge again. Merged in https://github.com/termux/termux-packages/commit/c114f860a01ea55dc1b5d08a468a010e357004db.

**Pull request merged**
